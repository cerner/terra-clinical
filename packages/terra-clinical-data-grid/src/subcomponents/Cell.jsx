import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames/bind';
import memoize from 'memoize-one';

import { KEYCODES } from '../utils/keycodes';

import styles from './Cell.module.scss';

const cx = classNames.bind(styles);

const propTypes = {
  /**
   * Accessible text provided for the Cell.
   */
  ariaLabel: PropTypes.string,
  /**
   * String identifier of the section in which the Cell will be rendered.
   */
  sectionId: PropTypes.string.isRequired,
  /**
   * String identifer of the row in which the Cell will be rendered.
   */
  rowId: PropTypes.string.isRequired,
  /**
   * String identifier of the column in which the Cell will be rendered.
   */
  columnId: PropTypes.string.isRequired,
  /**
   * String-formatted width that the Cell should be rendered as. Any valid css width value is supported (i.e. 200px, 3rem).
   */
  width: PropTypes.string.isRequired,
  /**
   * Boolean indicating whether the Cell is selectable.
   */
  isSelectable: PropTypes.bool,
  /**
   * Boolean indicating whether the Cell is actively selected.
   */
  isSelected: PropTypes.bool,
  /**
   * Function that will be called upon Cell selection. The `isSelectable` prop must be true for this function to be called.
   * Parameters: `onSelect(sectionId, rowId, columnId)`
   */
  onSelect: PropTypes.func,
  /**
   * Function that will be called upon the mouse entering the selectable region of the Cell. The `isSelectable` prop must be true for this function to be called.
   * Parameters: `onHoverStart(event)`
   */
  onHoverStart: PropTypes.func,
  /**
   * Function that will be called upon the mouse leaving the selectable region of the Cell. The `isSelectable` prop must be true for this function to be called.
   * Parameters: `onHoverEnd(event)`
   */
  onHoverEnd: PropTypes.func,
  /**
   * Content that will rendered within the Cell.
   */
  children: PropTypes.node,
  /**
   * Function that will be called with a ref to the Cell's selectable element. Parameters: `selectableRefCallback(selectableRef)`
   */
  selectableRefCallback: PropTypes.func,
};

class Cell extends React.Component {
  constructor(props) {
    super(props);

    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleTargetClick = this.handleTargetClick.bind(this);

    this.getCellStyles = memoize(this.getCellStyles);
  }

  /* eslint-disable class-methods-use-this */

  /**
   * This function is memoized in the constructor so that for a given width value, the same object reference will be returned.
   * This allows for repeat renders of the Cell to occur more efficiently if the width value has not changed between renders.
   */
  getCellStyles(width) {
    return {
      width,
    };
  }
  /* eslint-enable class-methods-use-this */

  handleKeyDown(event) {
    if (event.nativeEvent.keyCode === KEYCODES.ENTER || event.nativeEvent.keyCode === KEYCODES.SPACE) {
      const { onSelect } = this.props;

      if (onSelect) {
        event.preventDefault();
        onSelect(this.props.sectionId, this.props.rowId, this.props.columnId);
      }
    }
  }

  handleTargetClick() {
    const { onSelect } = this.props;

    if (onSelect) {
      onSelect(this.props.sectionId, this.props.rowId, this.props.columnId);
    }
  }

  render() {
    const {
      sectionId,
      rowId,
      columnId,
      isSelectable,
      isSelected,
      width,
      onSelect,
      children,
      selectableRefCallback,
      onHoverStart,
      onHoverEnd,
      ariaLabel,
      ...customProps
    } = this.props;

    return (
      <div
        {...customProps}
        className={cx(['container', customProps.className])}
        style={this.getCellStyles(width)}
        aria-selected={isSelected ? true : undefined}
      >
        <div
          role="button"
          className={cx(['content', { selectable: isSelectable, selected: isSelected }])}
          onClick={isSelectable ? this.handleTargetClick : undefined}
          onKeyDown={isSelectable ? this.handleKeyDown : undefined}
          onMouseEnter={onHoverStart}
          onMouseLeave={onHoverEnd}
          tabIndex={isSelectable ? '0' : undefined}
          ref={selectableRefCallback}
          aria-label={ariaLabel}
        >
          {children}
        </div>
      </div>
    );
  }
}

Cell.propTypes = propTypes;

export default Cell;
