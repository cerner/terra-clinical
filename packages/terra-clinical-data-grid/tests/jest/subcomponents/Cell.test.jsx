import React from 'react';
import Cell from '../../../src/subcomponents/Cell';

describe('Cell Snapshot Tests', () => {
  it('should render a Cell with minimal props', () => {
    const cell = shallow((
      <Cell
        sectionId="section-test"
        rowId="row-test"
        columnId="column-test"
        width="123px"
      />
    ));

    expect(cell).toMatchSnapshot();
  });

  it('should render a Cell with all provided props', () => {
    const cell = shallow((
      <Cell
        sectionId="section-test"
        rowId="row-test"
        columnId="column-test"
        width="123px"
        isSelectable
        isSelected
        onSelect={() => {}}
        onHoverStart={() => {}}
        onHoverEnd={() => {}}
        selectableRefCallback={() => {}}
        data-custom-attribute
        className="custom-class-name"
      >
        <div>Child here</div>
      </Cell>
    ));

    expect(cell).toMatchSnapshot();
  });
});
