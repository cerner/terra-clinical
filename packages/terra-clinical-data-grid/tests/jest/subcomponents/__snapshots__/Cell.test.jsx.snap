// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Snapshot Tests should render a Cell with all provided props 1`] = `
<div
  aria-selected={true}
  className="container custom-class-name"
  data-custom-attribute={true}
  style={
    Object {
      "width": "123px",
    }
  }
>
  <div
    className="content selectable selected"
    onClick={[Function]}
    onKeyDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    role="button"
    tabIndex="0"
  >
    <div>
      Child here
    </div>
  </div>
</div>
`;

exports[`Cell Snapshot Tests should render a Cell with minimal props 1`] = `
<div
  className="container"
  style={
    Object {
      "width": "123px",
    }
  }
>
  <div
    className="content"
    role="button"
  />
</div>
`;
